unix/vncpasswd/vncpasswd.man
'\" t
.\" ** The above line should force tbl to be a preprocessor **
.\" Man page for X vncpasswd
.\"
.\" Copyright (C) 2010, 2015, 2018, 2024 D. R. Commander
.\" Copyright (C) 2010 University Corporation for Atmospheric Research
.\" Copyright (C) 2005-2008 Sun Microsystems, Inc.
.\" Copyright (C) 2001-2003 Constantin Kaplinsky
.\" Copyright (C) 2000 Red Hat, Inc.
.\" Copyright (C) 1998 Marcus.Brinkmann@ruhr-uni-bochum.de
.\"
.\" You may distribute under the terms of the GNU General Public
.\" License as specified in the file LICENCE.TXT that comes with the
.\" TurboVNC distribution.
.\"
.TH vncpasswd 1 "July 2024" "" "TurboVNC"
.SH NAME
vncpasswd \- set passwords for TurboVNC sessions
.SH SYNOPSIS
\fBvncpasswd\fR [\-v] [\fIfile\fR]
.br
\fBvncpasswd\fR \-t [\-v]
.br
\fBvncpasswd\fR \-f
.br
\fBvncpasswd\fR \-o [\-v] [\-display \fIVNC-display\fR]
.br
\fBvncpasswd\fR \-c [\-display \fIVNC-display\fR]
.br
\fBvncpasswd\fR \-a \fIuser\fR [\-v] [\-display \fIVNC-display\fR]
.br
\fBvncpasswd\fR \-r \fIuser\fR [\-display \fIVNC-display\fR]
.br
.SH DESCRIPTION
The \fBvncpasswd\fR utility can be used to create and change
passwords for Xvnc (the TurboVNC virtual X server.)  \fBXvnc\fR uses these
passwords when started with the \fB\-rfbauth\fR command-line option
(or when started from the \fBvncserver\fR script.)

\fBvncpasswd\fR allows one to enter a primary VNC password and an
optional view-only password.  \fBXvnc\fR will ignore mouse and keyboard
input from viewers that authenticate with the view-only password.

The password filename defaults to \fB$HOME/.vnc/passwd\fR, unless the
\fB\-t\fR command-line option was used (see the OPTIONS section
below.)  The \fB$HOME/.vnc/\fR directory will be created if it does not
exist.

Each password has to be longer than five characters (unless the
\fB\-f\fR command-line option was used-- see its description below.)
Only the first eight characters are significant.  If the primary
password is too short, then vncpasswd will abort.  If the view-only
password is too short, then only the primary password will be saved.

Unless a filename was explicitly provided in the command line, then this
utility may perform certain sanity checks to prevent writing a
password file into some hazardous place.
.br
.SH ONE-TIME PASSWORD (OTP) SUPPORT
If the TurboVNC session is configured to support one-time passwords, then
\fBvncpasswd\fR can be used to generate new one-time passwords for both full
control and, optionally, view-only access.  A one-time password is valid for
one successful connection from a VNC viewer to the TurboVNC session.  For each
subsequent connection, a new one-time password must be generated.
\fBvncpasswd\fR can also be used to clear any one-time passwords that are
currently enabled for a TurboVNC session.
.br
.SH USER ACCESS CONTROL LIST SUPPORT
If the TurboVNC Server is configured to support user access control lists
(ACLs) for the PAM User/Password authentication method, then \fBvncpasswd\fR
can be used to add users to or remove users from a session's ACL.  Be careful
not to remove yourself from the list, or you will not be able to connect again
until you re-add yourself.
.br
.SH OPTIONS
.TP
\fB\-t\fR
Write passwords into \fB/tmp/$USER-vnc/passwd\fR, creating the
\fB/tmp/$USER-vnc/\fR directory if it does not exist and checking the
permissions on that directory (the mode must be 700.)  This option can
help to improve security if your home directory is shared on the
network (e.g. when using NFS.)
.TP
\fB\-f\fR
Filter mode.  Read plain-text passwords from stdin, and write encrypted
versions to stdout.  One or two passwords (full-control and view-only)
can be supplied in the input stream, with a new line terminating each password.
Note that in filter mode, short or even empty passwords will be
silently accepted.
.TP
\fB\-v\fR
Also prompt for a view-only password, or generate a one-time view-only password.

If adding a user to the user ACL, then the \fB\-v\fR option specifies that the
user should be granted view-only access instead of full control.
.TP
\fB\-display\fR \fIVNC-display\fR
This option specifies the X display of the TurboVNC session to use when
setting/clearing one-time passwords or adding users to/removing users from the
ACL.  If \fIVNC-display\fR is not specified, then it defaults to the value of
the \fBDISPLAY\fR environment variable.
.TP
\fB\-o\fR
Generate a one-time full control password for \fIVNC-display\fR.
.TP
\fB\-c\fR
Clear any one-time passwords on \fIVNC-display\fR.
.TP
\fB\-a\fR \fIuser\fR
Add \fIuser\fR to \fIVNC-display\fR's user ACL.  The \fB\-v\fR option can also
be specified in order to restrict this user's access to view-only.
.TP
\fB\-r\fR \fIuser\fR
Remove \fIuser\fR from \fIVNC-display\fR's user ACL.
.SH EXIT CODES
If at least one password was saved successfully, or if the one-time password
or user ACL operation succeeded, then \fBvncpasswd\fR will
exit with status 0.  Otherwise, it will exit with status 1.
.SH SEE ALSO
\fBvncserver\fR(1), \fBXvnc\fR(1), \fBvncviewer\fR(1),
\fBvncconnect\fR(1)
.SH AUTHORS
VNC was originally developed at AT&T Laboratories Cambridge.  TightVNC
additions were implemented by Constantin Kaplinsky.  TurboVNC, based
on TightVNC, is provided by The VirtualGL Project.  Many other people
participated in development, testing and support.

\fBMan page authors:\fR
.br
Marcus Brinkmann <Marcus.Brinkmann@ruhr-uni-bochum.de>
.br
Tim Waugh <twaugh@redhat.com>
.br
Constantin Kaplinsky <const@tightvnc.com>
.br
D. R. Commander <information@turbovnc.org>
.br
Craig Ruff <cruff@ucar.edu>


ChangeLog.md
3.2 Evolving
============

### Significant changes relative to 3.1.1:

1. Since all supported TurboVNC host platforms now contain new enough libraries
to build xorg-server 1.20.x, the TurboVNC Server is now dynamically linked
against the system-supplied versions of these libraries rather than statically
linked against in-tree versions.  The `TVNC_SYSTEMLIBS` and `TVNC_SYSTEMX11`
CMake variables have been removed, and the build system now behaves as if those
variables are always on.  A new CMake variable (`TVNC_INTELZLIB`) can be used
on x86 platforms to disable the in-tree SIMD-accelerated Intel zlib
implementation and build against the system-supplied zlib implementation.

2. The TurboVNC Server now supports the DRI3 X extension when using open source
GPU drivers.  This enables GPU acceleration in a TurboVNC session without
VirtualGL, although the performance will be better with VirtualGL.  Refer to
the description of the `-drinode` option in the Xvnc man page for more details.

3. The default X startup script (`xstartup.turbovnc`) now throws an error,
rather than trying to execute **xinitrc** or twm, if a session desktop file for
the default window manager cannot be found.


3.1.2
=====

### Significant changes relative to 3.1.1:

1. The TurboVNC Server now assigns an ordinal ID to every VNC viewer when the
viewer connects, and when possible, log entries related to a specific viewer
are prefixed by the viewer's ID.  The TurboVNC Server now also logs the total
number of simultaneously connected viewers.

2. The TurboVNC Viewer now sends horizontal scroll wheel events to the VNC
server.  (These events can be generated with horizontal scroll gestures on a
trackpad or, with certain mice, by side-clicking the scroll wheel.)

3. By default, the TurboVNC Server now limits the amount of time that it will
wait for a new pointer event from a connected viewer that is dragging the mouse
(and thus has exclusive control over the pointer.)  This prevents other viewers
connected to the same session from being locked out of pointer control
indefinitely if a viewer's network connection drops while it is dragging the
mouse.  A new Xvnc command-line option (`-pointerlocktimeout`) can be used to
specify the time limit.

4. The RPM and DEB packages generated by the TurboVNC build/packaging system
now include a polkit rules file that prevents various authentication dialogs
("Authentication is required to create a color managed device", "Authentication
is required to access the PC/SC daemon", "Authentication is required to refresh
the system repositories") from popping up when using the GNOME window manager
with the TurboVNC Server on Ubuntu 23.10 and later (if the `polkitd-pkla`
package is not installed) and on RHEL 7 and Fedora 19 and later (if the
`polkit-pkla-compat` package is not installed.)

5. The default X startup script (`xstartup.turbovnc`) now throws an error,
rather than trying to execute **xinitrc** or twm, if a window manager is
specified and the session desktop file for the window manager cannot be found.

6. Added a new parameter (`ExtSSHCommand`) to the TurboVNC Viewer that can be
used to specify the external SSH client command.  In `Via`/`Tunnel` SSH
command-line templates, including the default ones, `%S` is now replaced with
the value of this new parameter.


3.1.1
=====

### Significant changes relative to 3.1:

1. By default, each instance of the Linux TurboVNC Server now listens on the
abstract Unix domain socket, in addition to the pathname Unix domain socket
(under **/tmp/.X11-unix**), associated with its X display number.  This
prevents recent versions of GDM, when configured with `WaylandEnable=false`,
from attempting to use Display :1 for the local session if a TurboVNC session
is already using Display :1.  The previous behavior can be restored by passing
`-nolisten local` to `vncserver` or adding `-nolisten local` to the
`$serverArgs` variable in **turbovncserver.conf**.

2. The `vncserver` script now checks whether the abstract Unix domain socket
associated with an X display number is in use before assuming that the display
number is available.

3. Fixed an issue in the Windows TurboVNC Viewer whereby an F10 key press,
followed by an F10 key release, caused the keyboard focus to be redirected to
the system menu, and subsequent keystrokes were consumed by the system menu
until F10, left Alt, or Esc was pressed to dismiss the menu.

4. Fixed an issue whereby GTK applications (including the GNOME window manager)
running in a TurboVNC session attempted to display to a local Wayland session
if one was active.


3.1
===

### Significant changes relative to 3.1 beta2:

1. Improved the TurboVNC Viewer's handling of SSH usernames in the following
ways:

     - Fixed a regression introduced in 3.1 beta1[3] whereby the SSH username
was ignored if it was specified in the `Server` parameter or if it was
specified in the TurboVNC Viewer Options dialog without also specifying the
gateway host.
     - Fixed an issue whereby the SSH username was not saved and restored if it
was specified in the TurboVNC Viewer Options dialog without also specifying the
gateway host.
     - Fixed an issue whereby the SSH username was ignored if it was specified
in the `Server` or `Via` parameter in **~/.vnc/default.turbovnc**.
     - Added a new parameter (`SSHUser`) that can optionally be used to specify
the SSH username.  This parameter is set automatically from an SSH username
specified in the `Server` or `Via` parameter, or it can be set manually.
     - To better emulate the behavior of OpenSSH, the TurboVNC Viewer's
built-in SSH client now allows an SSH username specified on the command line or
in a connection info file to override an SSH username specified in the OpenSSH
config file.
     - The `LocalUsernameLC` parameter now affects the SSH username if the SSH
username is unspecified.

2. The TurboVNC Server now includes various security fixes (CVE-2022-2319,

unix/vncpasswd/vncpasswd.c
/* Copyright (C) 2010, 2019, 2021 D. R. Commander.  All Rights Reserved.
 * Copyright (C) 2010 University Corporation for Atmospheric Research.
 *                    All Rights Reserved.
 * Copyright (C) 2002-2003 Constantin Kaplinsky.  All Rights Reserved.
 * Copyright (C) 1999 AT&T Laboratories Cambridge.  All Rights Reserved.
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
 * USA.
 */

/*
 *  vncpasswd:  A standalone program that gets and verifies a password,
 *              encrypts it, and stores it to a file.  Optionally, it does
 *              the same for a second (view-only) password.  Always ignore
 *              anything after 8 characters, since this is what Solaris
 *              getpass() does anyway.
 */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>
#include "vncauth.h"

#include <X11/Xlib.h>
#include <X11/Xatom.h>

#ifdef UseDevUrandom
#define URANDOM_PATH "/dev/urandom"

#include <stdint.h>
#include <fcntl.h>
#else
#include <sys/time.h>
#endif


static void usage(char *argv[]);
static char *getenv_safe(char *name, size_t maxlen);
static void mkdir_and_check(char *dirname, int be_strict);
static int read_password(char *result);
static int ask_password(char *result);

int alsoView;

int otp;
int otpClear;
char *displayname;


static int DoOTP(void)
{
  unsigned int full;
  unsigned int view = 0;
  Display *dpy;
  Atom prop;
  int len;
  char buf[MAXPWLEN + 1];
  char bytes[MAXPWLEN * 2];
#ifdef UseDevUrandom
  int fd;
#endif

  if ((dpy = XOpenDisplay(displayname)) == NULL) {
    fprintf(stderr, "unable to open display \"%s\"\n",
            XDisplayName(displayname));
    return(1);
  }

  prop = XInternAtom(dpy, "VNC_OTP", True);
  if (prop == None) {
    fprintf(stderr, "The X display \"%s\" does not support VNC one-time passwords\n",
            XDisplayName(displayname));
    return(1);
  }

  if (otpClear) {
    len = 0;

  } else {
#ifdef UseDevUrandom
    if ((fd = open(URANDOM_PATH, O_RDONLY, 0)) < 0) {
      perror(URANDOM_PATH);
      return(1);
    }

    if (
      (read(fd, &full, sizeof(full)) != sizeof(full)) ||
      (alsoView && (read(fd, &view, sizeof(view)) != sizeof(view)))
    ) {
      fprintf(stderr, "Could not read random number from %s\n", URANDOM_PATH);
      return(1);
    }

    close(fd);

#else
    struct timeval now;

    gettimeofday(&now, NULL);
    srandom((unsigned int)now.tv_sec + (unsigned int)now.tv_usec);
    full = random();
    if (alsoView)
      view = random();
#endif

    snprintf(buf, sizeof(buf), "%08u", full);
    memcpy(&bytes[0], buf, MAXPWLEN);
    fprintf(stderr, "Full control one-time password: %.*s\n", MAXPWLEN,
            &bytes[0]);
    len = MAXPWLEN;
    if (alsoView) {
      snprintf(buf, sizeof(buf), "%08u", view);
      memcpy(&bytes[MAXPWLEN], buf, MAXPWLEN);
      fprintf(stderr, "View-only one-time password: %.*s\n", MAXPWLEN,
              &bytes[MAXPWLEN]);
      len = MAXPWLEN * 2;
    }
  }

#ifdef notdef
  if (len >= MAXPWLEN)
    vncEncryptPasswd(&bytes[0], &bytes[0]);

  if (len == MAXPWLEN * 2)
    vncEncryptPasswd(&bytes[MAXPWLEN], &bytes[MAXPWLEN]);
#endif

  XChangeProperty(dpy, DefaultRootWindow(dpy), prop, XA_STRING, 8,
                  PropModeReplace, (unsigned char *)bytes, len);
  memset(bytes, 0, sizeof(bytes));
  XCloseDisplay(dpy);
  return 0;
}


int addUser;
int userList;
char *user;

#define MAXUSERLEN  63

static int DoUserList(void)
{
  Display *dpy;
  Atom prop;
  int len;
  char bytes[MAXUSERLEN + 1];

  if ((user == NULL) || ((len = strlen(user)) == 0)) {
    fprintf(stderr, "missing the username!");
    return 1;
  }

  if (len > MAXUSERLEN) {
    fprintf(stderr, "username is too large");
    return 1;
  }

  if ((dpy = XOpenDisplay(displayname)) == NULL) {
    fprintf(stderr, "unable to open display \"%s\"\n",
            XDisplayName(displayname));
    return 1;
  }

  prop = XInternAtom(dpy, "VNC_ACL", True);
  if (prop == None) {
    fprintf(stderr, "The X server \"%s\" does not support VNC user access control lists\n",
            XDisplayName(displayname));
    return 1;
  }

  bytes[0] = addUser | (alsoView ? 0x10 : 0x00);
  memcpy(&bytes[1], user, len);
  XChangeProperty(dpy, DefaultRootWindow(dpy), prop, XA_STRING, 8,
                  PropModeReplace, (unsigned char *)bytes, len + 1);
  XCloseDisplay(dpy);
  return 0;
}


int main(int argc, char *argv[])
{
  int read_from_stdin = 0;
  int make_directory = 1;
  int check_strictly = 0;
  char passwd1[9];
  char passwd2[9];
  char *passwd2_ptr;
  char yesno[2];
  char passwdDir[256];
  char passwdFile[263];
  int i;

  sprintf(passwdDir, "%s/.vnc", getenv_safe("HOME", 240));
  sprintf(passwdFile, "%s/passwd", passwdDir);

  for (i = 1; i < argc; i++) {
    if (argv[i][0] != '-')
      break;

    switch (argv[i][1]) {
      case 'd':
        if (strcmp("-display", argv[i]))
          usage(argv);

        if (++i >= argc)
          usage(argv);

        displayname = argv[i];
        break;

      case 'c':
        make_directory = 0;
        otp = 1;  otpClear = 1;
        break;

      case 'o':
        make_directory = 0;
        otp = 1;
        break;

      case 'a':
        if (++i >= argc)
          usage(argv);

        user = argv[i];
        make_directory = 0;
        userList = 1;
        addUser = 1;
        break;

      case 'r':
        if (++i >= argc)
          usage(argv);

        user = argv[i];
        make_directory = 0;
        userList = 1;
        addUser = 0;
        break;

      case 'f':
        strcpy(passwdFile, "-");
        read_from_stdin = 1;
        make_directory = 0;
        check_strictly = 0;
        break;

      case 't':
        sprintf(passwdDir, "/tmp/%s-vnc", getenv_safe("USER", 32));
        sprintf(passwdFile, "%s/passwd", passwdDir);
        read_from_stdin = 0;
        make_directory = 1;
        check_strictly = 1;
        break;

      case 'v':
        alsoView = 1;
        break;

      default:
        usage(argv);
        break;
    }
  }

  if (otp) {
    if (read_from_stdin) {
      fprintf(stderr, "Error: -f is incompatible with -o\n");
      exit(1);
    }

    if (!strncmp(passwdDir, "/tmp", 4)) {
      fprintf(stderr, "Error: -t is incompatible with -o\n");
      exit(1);
    }

    if (userList) {
      fprintf(stderr, "Error: -a and -r are incompatible with -o\n");
      exit(1);
    }

    exit(DoOTP());
  }

  if (userList) {
    if (read_from_stdin) {
      fprintf(stderr, "Error: -f is incompatible with -a and -r\n");
      exit(1);
    }

    if (!strncmp(passwdDir, "/tmp", 4)) {
      fprintf(stderr, "Error: -t is incompatible with -a and -r\n");
      exit(1);
    }

    if (otp) {
      fprintf(stderr, "Error: -o is incompatible with -a and -r\n");
      exit(1);
    }

    exit(DoUserList());
  }

  if (i != argc) {
    if (strlen(argv[i]) > sizeof(passwdFile) - 1) {
      fprintf(stderr, "Error: file name too long\n");
      exit(1);
    }
    strcpy(passwdFile, argv[i]);
    if (read_from_stdin) {
      fprintf(stderr, "Error: cannot specify filename with -f\n");
      exit(1);
    }
    read_from_stdin = 0;
    make_directory = 0;
    check_strictly = 0;
  }

  if (make_directory) {
    fprintf(stderr, "Using password file %s\n", passwdFile);
    mkdir_and_check(passwdDir, check_strictly);
  }

  passwd2_ptr = NULL;

  if (read_from_stdin) {

    /* Read one or two passwords from stdin */
    if (!read_password(passwd1)) {
      fprintf(stderr, "Could not read password\n");
      exit(1);
    }
    if (read_password(passwd2)) {
      passwd2_ptr = passwd2;
    }

  } else {

    /* Ask the primary (full-control) password. */
    if (!ask_password(passwd1)) {
      exit(1);
    }

    /* Optionally, ask the second (view-only) password. */
    if (alsoView) {
      fprintf(stderr, "Enter the view-only password\n");

    } else {
      /* FIXME: Is it correct to read from stdin here? */
      fprintf(stderr, "Would you like to enter a view-only password (y/n)? ");
      if (fgets(yesno, 2, stdin) != NULL && strchr("Yy", yesno[0]) != NULL)
        alsoView = 1;
    }

    if (alsoView && ask_password(passwd2)) {
      passwd2_ptr = passwd2;
    }
  }

  /* Actually write the passwords. */
  if (!vncEncryptAndStorePasswd2(passwd1, passwd2_ptr, passwdFile)) {
    memset(passwd1, 0, strlen(passwd1));
    memset(passwd2, 0, strlen(passwd2));
    fprintf(stderr, "Cannot write password file %s\n", passwdFile);
    exit(1);
  }

  /* Zero the memory. */
  memset(passwd1, 0, strlen(passwd1));
  memset(passwd2, 0, strlen(passwd2));
  return 0;
}


static void usage(char *argv[])
{
  fprintf(stderr, "usage: %s [-v] [FILE]\n", argv[0]);
  fprintf(stderr, "       %s -f\n", argv[0]);
  fprintf(stderr, "       %s -t [-v]\n", argv[0]);
  fprintf(stderr, "       %s -o [-v] [-display VNC-DISPLAY]\n", argv[0]);
  fprintf(stderr, "       %s -c [-display VNC-DISPLAY]\n", argv[0]);
  fprintf(stderr, "       %s -a USER [-v] [-display VNC-DISPLAY]\n", argv[0]);
  fprintf(stderr, "       %s -r USER [-display VNC-DISPLAY]\n", argv[0]);
  exit(1);
}


static char *getenv_safe(char *name, size_t maxlen)
{
  char *result;

  result = getenv(name);
  if (result == NULL) {
    fprintf(stderr, "Error: no %s environment variable\n", name);
    exit(1);
  }
  if (strlen(result) > maxlen) {
    fprintf(stderr, "Error: %s environment variable string too long\n", name);
    exit(1);
  }
  return result;
}


/*
 * Check if the specified vnc directory exists, create it if
 * necessary, and perform a number of sanity checks.
 */

static void mkdir_and_check(char *dirname, int be_strict)
{
  struct stat stbuf;

  if (lstat(dirname, &stbuf) != 0) {
    if (errno != ENOENT) {
      fprintf(stderr, "lstat() failed for %s: %s\n", dirname, strerror(errno));
      exit(1);
    }
    fprintf(stderr, "VNC directory %s does not exist, creating.\n", dirname);
    if (mkdir(dirname, S_IRWXU) == -1) {
      fprintf(stderr, "Error creating directory %s: %s\n", dirname,
              strerror(errno));
      exit(1);
    }
  }

  if (lstat(dirname, &stbuf) != 0) {
    fprintf(stderr, "Error in lstat() for %s: %s\n", dirname, strerror(errno));
    exit(1);
  }
  if (!S_ISDIR(stbuf.st_mode)) {
    fprintf(stderr, "Error: %s is not a directory\n", dirname);
    exit(1);
  }
  if (stbuf.st_uid != getuid()) {
    fprintf(stderr, "Error: bad ownership on %s\n", dirname);
    exit(1);
  }
  if (be_strict && ((S_IRWXG | S_IRWXO) & stbuf.st_mode)) {
    fprintf(stderr, "Error: bad access modes on %s\n", dirname);
    exit(1);
  }
}


/*
 * Read a password from stdin.  The password is terminated either by an
 * end of line, or by the end of stdin data.  Return 1 on success, 0 on
 * error. On success, the password will be stored in the specified
 * 9-byte buffer.
 */

static int read_password(char *result)
{
  char passwd[256];
  char *ptr;

  /* Try to read the password. */
  if (fgets(passwd, 256, stdin) == NULL)
    return 0;

  /* Remove the newline if present. */
  ptr = strchr(passwd, '\n');
  if (ptr != NULL)
    *ptr = '\0';

  /* Truncate if necessary. */
  if (strlen(passwd) > 8) {
    memset(passwd + 8, 0, strlen(passwd) - 8);
    fprintf(stderr, "Warning: password truncated to the length of 8.\n");
  }

  /* Save the password and zero our copies. */
  strcpy(result, passwd);
  memset(passwd, 0, strlen(passwd));

  return 1;
}


/*
 * Ask a password, check its length and ask to confirm it once more.
 * Return 1 on success, 0 on error.  On success, the password will be
 * stored in the specified 9-byte buffer.
 */

static int ask_password(char *result)
{
  char *passwd;
  char passwd_copy[9];

  while (1) {
    passwd = getpass("Password: ");
    if (!passwd) {
      fprintf(stderr, "Can't get password: not a tty?\n");
      return 0;
    }
    if (strlen(passwd) < 6) {
      fprintf(stderr, "Password too short\n");
      return 0;
    }
    if (strlen(passwd) > 8) {
      memset(passwd + 8, 0, strlen(passwd) - 8);
      fprintf(stderr, "Warning: password truncated to the length of 8.\n");
    }

    strcpy(passwd_copy, passwd);

    passwd = getpass("Verify:   ");
    if (strlen(passwd) > 8)
      memset(passwd + 8, 0, strlen(passwd) - 8);

    if (strcmp(passwd, passwd_copy) == 0)
      break;                    /* success */

    fprintf(stderr, "Passwords do not match. Please try again.\n\n");
  }

  /* Save the password and zero our copies. */
  strcpy(result, passwd);
  memset(passwd, 0, strlen(passwd));
  memset(passwd_copy, 0, strlen(passwd_copy));

  return 1;
}
